File: tema4.js
const util = require('util');

const fs = require('fs');
precision = 1;

for (let i = 0; i < 10; i++) {
  precision /= 10;
}
// File content
let file_content = fs.readFileSync('m_rar_2018_5.txt', 'utf8');

//
let file_splited = file_content.split('\r\n').filter(e => e != '');
let n = parseInt(file_splited[0]);

file_splited.splice(0, 1);

// B vector
let b = [];

for (let i = 0; i < n; i++) {
  b.push(file_splited[i]);
}

// Matrix
let a = {};
for (let i = n; i < file_splited.length; i++) {
  tmp = file_splited[i].split(/[ ,]+/);
  if (!a[tmp[1]]) {
    a[tmp[1]] = {};
  }

  a[tmp[1]][tmp[2]] = Number.parseFloat(tmp[0]);
}

// Verific daca diagonala are element null
sw = false;
for (let i = 0; i < n; i++) {
  if (typeof a[i][i] === 'undefined') {
    console.log(`[MSG]: ELEMENT NULL PE DIAGOLANA. POZITIA: ${i}`);
    sw = true;
    break;
  }
}

equal = (previos, actual) => {
  for (i = 0; i < previos.length; i++) {
    pr = previos[i];
    nx = actual[i];
    if (
      Math.abs(previos[i] - parseInt(previos[i])) <= precision ||
      Math.abs(parseInt(previos[i]) + 1 - previos[i]) <= precision
    ) {
      pr = parseInt(previos[i]);
    }
    if (
      Math.abs(actual[i] - parseInt(actual[i])) <= precision ||
      Math.abs(parseInt(actual[i]) + 1 - actual[i]) <= precision
    ) {
      nx = parseInt(actual[i]);
    }
    if (pr != nx) return false;
  }
  return true;
};

if (!sw) {
  console.log('[MSG]: ELEMENTE NENULE PE DIAGONALA');

  aCopy = Object.assign({}, a);
  calculatedX = [];
  for (i = 0; i < n; i++) {
    calculatedX.push(Math.floor(Math.random() * 2));
  }
  keys = [];

  for (key of Object.keys(a)) {
    keys[key] = Object.keys(a[key]);
  }

  // previous = calculatedX.map(e => e + 1);
  iterations = 0;
  broken = false;
  do {
    previous = calculatedX.slice();
    iterations++;
    // console.log(previous);

    for (j = 0; j < n; j++) {
      tmpX = 0;
      for (key of keys[j]) {
        if (+key === j) continue;
        // t = a[j][key] * calculatedX[+key];
        // if (!t && t !== 0) {
        //     broken = true;
        //     break;
        // }
        tmpX -= a[j][key] * calculatedX[+key];
      }

      tmpX += +b[j];
      tmpX /= a[j][j];
      calculatedX[j] = tmpX;
    }
  } while (!equal(previous, calculatedX) && iterations < 50000);

  for (i = 0; i < calculatedX.length; i++) {
    fs.appendFile(
      'test.txt',
      `${calculatedX[i]}\n${previous[i]}\n\n`,
      'utf8',
      err => err
    );
  }

  console.log(calculatedX);
  console.log(iterations);

  finalResult = 0;
  for (i = 0; i < n; i++) {
    s = 0;
    for (key of keys[i]) {
      s += a[i][key] * calculatedX[key];
    }

    s -= b[i];
    finalResult += s * s;
  }

  console.log(finalResult);
}

File: tema4.py
import copy

for fisier in [1, 2, 3, 4]:
    f = open('m_rar_2018_' + str(fisier) + '.txt', 'r')
    content_file = f.read()
    f.close()

    precision = 1
    for i in range(0, 10):
        precision /= 10

    content_file_array = content_file.split('\n')
    content_file_array = [a for a in content_file_array if a != '']

    length = int(content_file_array[0])
    content_file_array.pop(0)

    B = [float(content_file_array[i]) for i in range(0, length)]

    elements = [[] for i in range(0, length)]
    index = [[] for i in range(0, length)]
    diagonal = []

    for i in range(length, len(content_file_array)):
        tmp = content_file_array[i].split(', ')
        tmp = [float(i) for i in tmp]
        tmp[1] = int(tmp[1])
        tmp[2] = int(tmp[2])

        if (tmp[1] == tmp[2]):
            diagonal += [tmp[0]]

        elements[tmp[1]] += [tmp[0]]
        index[tmp[1]] += [tmp[2]]

    # elements = [[12, 3, -5], [1, 5, 3], [3, 7, 13], [4, -5, 33]]
    # index = [[0, 2, 3], [0, 1, 2], [0, 2, 3], [1, 2, 3]]
    # B = [1, 28, 76, 44]

    diagonal_has_0_elements = False
    for i in range(0, len(diagonal)):
        if abs(diagonal[i]) < precision:
            diagonal_has_0_elements = True
            break

    def equal(previous, nxt):
        for i in range(0, len(previous)):
            p = previous[i]
            n = nxt[i]

            # print(p, n)

            # if abs(p - n) > precision:
            if p != n:
                # print(previous, nxt)
                return False
        return True

    def getIndex(i):
        for j in range(0, len(index[i])):
            if index[i][j] == i:
                return j
        return 0

    if not diagonal_has_0_elements:
        X = [0 for i in B]
        X_prev = [i + 1 for i in X]
        counter = 0

        while not equal(X_prev, X) and counter < 250000:
            X_prev = copy.copy(X)
            for i in range(0, len(elements)):
                tmp = B[i]

                for j in range(0, len(index[i])):
                    if i == index[i][j]:
                        continue

                    tmp -= elements[i][j] * X[index[i][j]]
                    # print("elements", elements[i][j], X[index[i][j]])

                tmp /= elements[i][getIndex(i)]
                # print("tmp", tmp, B[i])
                # print(X[i], B[i], elements[i][getIndex(i)], tmp)
                X[i] = tmp

            counter += 1

        print("FISIER ", fisier)
        print("number of iterations: ", counter)
        finalResult = 0
        for i in range(0, len(elements)):
            s = 0
            for j in range(0, len(index[i])):
                s += elements[i][j] * X[index[i][j]]
            s -= B[i]
            finalResult += s * s

        print("| Ax -  b | ", finalResult)
        print(' ')
